services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceControllerService.java

```diff
@@ -84,13 +90,21 @@ public VeniceControllerService(
     }
     // The admin consumer needs to use VeniceHelixAdmin to update Zookeeper directly
     consumerServicesByClusters = new HashMap<>(multiClusterConfigs.getClusters().size());
+    PubSubTopicRepository pubSubTopicRepository = new PubSubTopicRepository();
+    KafkaValueSerializer kafkaValueSerializer = new OptimizedKafkaValueSerializer();
+    kafkaMessageEnvelopeSchemaReader.ifPresent(kafkaValueSerializer::setSchemaReader);
+    KafkaPubSubMessageDeserializer pubSubMessageDeserializer = new KafkaPubSubMessageDeserializer(
+        kafkaValueSerializer,
+        new LandFillObjectPool<>(KafkaMessageEnvelope::new),
+        new LandFillObjectPool<>(KafkaMessageEnvelope::new));
```

Comment:
Are `PubSubTopicRepository` and `ObjectPool` supposed to be singleton instances to avoid duplication?

Reply:
Yeah, but this is the only place in the controller where they are instantiated. I prefer avoiding the "singleton as static instance" pattern since that is weird in our integration tests where many different processes should not share static state.

Reply:
Yeah, but this is the only place in the controller where they are instantiated. I prefer avoiding the "singleton as static instance" pattern since that is weird in our integration tests where many different processes should not share static state.

Reply:
SG.